### PR TITLE
Rec: shared throttle map

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

Move throttle map from a thread local to shared. This map is only consulted when going auth on the net. Information collected by a thread is useful to other threads to avoid double work finding out an auth needs to be throttled.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [X] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/syncres.cc [hunk 1]
```diff
@@ -3126,12 +3231,12 @@ vector<ComboAddress> SyncRes::retrieveAddressesForNS(const std::string& prefix,
 
 bool SyncRes::throttledOrBlocked(const std::string& prefix, const ComboAddress& remoteIP, const DNSName& qname, const QType qtype, bool pierceDontQuery)
 {
-  if(t_sstorage.throttle.shouldThrottle(d_now.tv_sec, std::make_tuple(remoteIP, g_rootdnsname, 0))) {
+  if (s_throttle.lock()->shouldThrottle(d_now.tv_sec, std::make_tuple(remoteIP, g_rootdnsname, 0))) {
```
Comment (rgacogne) [inline pdns/syncres.cc:3234]: I wonder if it would make sense to use the wrappers from the SyncRes class instead?

**Review by rgacogne — APPROVED** [review]
Code looks good! We probably need to document that change, though?

File: pdns/syncres.cc (hunk 1 above)
Comment (omoerbeek) [reply pdns/syncres.cc]: yep, also did it for the other cases, which makes the code a little bit easier to read.

File: pdns/syncres.cc [hunk 2]
```diff
@@ -4950,15 +5060,15 @@ bool SyncRes::doResolveAtThisIP(const std::string& prefix, const DNSName& qname,
       if (s_serverdownmaxfails > 0 && (auth != g_rootdnsname) && s_fails.lock()->incr(remoteIP, d_now) >= s_serverdownmaxfails) {
         LOG(prefix<<qname<<": Max fails reached resolving on "<< remoteIP.toString() <<". Going full throttle for "<< s_serverdownthrottletime <<" seconds" <<endl);
         // mark server as down
-        t_sstorage.throttle.throttle(d_now.tv_sec, std::make_tuple(remoteIP, g_rootdnsname, 0), s_serverdownthrottletime, 10000);
+        doThrottle(d_now.tv_sec, remoteIP, g_rootdnsname, 0, s_serverdownthrottletime, 10000);
```
Comment (rgacogne) [inline pdns/syncres.cc:5063]:
```suggestion
        doThrottle(d_now.tv_sec, remoteIP, s_serverdownthrottletime, 10000);
```

**Review by rgacogne — APPROVED** [review]
Looks good, one cosmetic suggestion.

**Pull request merged**